src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7363,9 +7363,19 @@ private BoundExpression ConvertToArrayIndex(BoundExpression index, SyntaxNode no
 
             var result =
                 TryImplicitConversionToArrayIndex(index, SpecialType.System_Int32, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt32, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_Int64, node, diagnostics) ??
-                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt64, node, diagnostics);
+                TryImplicitConversionToArrayIndex(index, SpecialType.System_UInt32, node, diagnostics);
+
+            if (result is null && allowNIntAndNUInt)
+            {
+                result = TryImplicitConversionToArrayIndex(index, SpecialType.System_IntPtr, node, diagnostics) ??
+                         TryImplicitConversionToArrayIndex(index, SpecialType.System_UIntPtr, node, diagnostics);
```

Comment:
>SpecialType.System_UIntPtr [](start = 66, length = 26)

ECMA-335 says: "index (of type int32 or
native int)" Nothing about unsigned native int.

Reply:
That's true, but the current JIT does support it. It's possibly worth checking if this is just one of the many places ECMA-335 is out of sync with what the current JIT allows.

Roslyn's own logic is also slightly questionable here:
<https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHAZgAJ9KBhSgb1MpeqrQDsZKBZACk4wA2gF1KEJJQCugypwAmcAB4BKSgF4AfOKELlIgNzNWxlhTldeArqPGSpGAPYcA5hcWqN2iLo4fDpAC+QA>

That is, you can see that the `uint32` is passed in "as-is" to `ldelem`, where if it was "strictly" following the spec it would be doing `conv.ovf.i4.un` or `conv.ovf.i.un`